File: projeto_biblioteca/models/nlivro.rb
require 'json'
require_relative 'crud'

class Livro < Crud
  attr_accessor :id, :titulo, :autor, :genero, :num_paginas, :ano_publicacao, :disponibilidade, :qtd_total, :qtd_disponivel

  def initialize(id, titulo, autor, genero, num_paginas, ano_publicacao, disponibilidade, qtd_total, qtd_disponivel)
    super('livros.json')
    @id = id
    @titulo = titulo
    @autor = autor
    @genero = genero
    @num_paginas = num_paginas
    @ano_publicacao = ano_publicacao
    @disponibilidade = disponibilidade
    @qtd_total = qtd_total
    @qtd_disponivel = qtd_disponivel
  end

  def abrir
    listar
  end

  def salvar
    dados = listar
    dados << { id: @id, titulo: @titulo, autor: @autor, genero: @genero, num_paginas: @num_paginas, ano_publicacao: @ano_publicacao, disponibilidade: @disponibilidade, qtd_total: @qtd_total, qtd_disponivel: @qtd_disponivel }
    File.write(@caminho_arquivo, JSON.pretty_generate(dados))
  end

  def listar
    if File.exist?(@caminho_arquivo)
      arquivo = File.read(@caminho_arquivo)
      JSON.parse(arquivo)
    else
      []
    end
  end
end

File: projeto_biblioteca/models/nlivro.js
const CRUD = require('./CRUD');
const fs = require('fs');

class Livro extends CRUD {
    constructor(id, titulo, autor, genero, num_paginas, ano_publicacao, disponibilidade, qtd_total, qtd_disponivel) {
        super('livros.json');
        this.id = id;
        this.titulo = titulo;
        this.autor = autor;
        this.genero = genero;
        this.num_paginas = num_paginas;
        this.ano_publicacao = ano_publicacao;
        this.disponibilidade = disponibilidade;
        this.qtd_total = qtd_total;
        this.qtd_disponivel = qtd_disponivel;
    }

    abrir() {
        return this.listar();
    }

    salvar() {
        const dados = this.listar();
        dados.push(this);
        fs.writeFileSync(this.caminhoArquivo, JSON.stringify(dados, null, 2));
    }
}

module.exports = Livro;
